### ISSUE TITLE
Improve error message when adding ServiceEndpoint to service builder for a service type which hasn't been added

### ISSUE THREAD

**Issue by manuelelucchi** [CoreWCF/CoreWCF#437]
Hello,
while configuring a NetTcp binding using CoreWCF I get this really not useful exception. I tracked it and it should be thrown by aspnetcore. Since this is pretty much equal to the example you provide, I'm not sure which is the problem, so if you have any idea I would be happy to listen :D. Below I uploaded the startup, the webhost, the exception and even the stacktrace (TConnector is a generic, but in this case it's a class that implements IBqConnector with and empty constructor)
Thank you,
Manuele

![image](https://user-images.githubusercontent.com/17146198/132362865-f6259e39-5f65-4647-8e33-3998937cfe05.png)

![image](https://user-images.githubusercontent.com/17146198/132362762-c02d328b-cbfd-478a-af8f-190a71402543.png)

![image](https://user-images.githubusercontent.com/17146198/132362942-2b732dd9-4a38-4156-a4ca-0dd1e1cbce7c.png)

![image](https://user-images.githubusercontent.com/17146198/132363021-b7524670-89f8-4644-9390-f6061c57c049.png)

**Comment by mconnew** [comment]
You're missing `builder.AddService<TConnector>();` before the `builder.AddServiceEndpoint` so CoreWCF doesn't know about the service type TService when trying to add an endpoint for it. The exception message could be improved to explain what the problem is here.

**Comment by manuelelucchi** [comment]
Yeah, now it works, thanks!

**Comment by mconnew** [comment]
I've edited the issue title to use this issue to track improvements to the error message

**Comment by Kralizek** [comment]
@mconnew please don't take the following nothing more as a constructive feedback. If it sounds anything but that, blame it on me not being English native.

Is there any particular reason why both the service and its endpoints are applied directly to the builder? Probably a nested approach would have helped avoid cases like this.

I was imagining something like

```csharp
services.AddServiceModel(builder => {
    builder.AddService<CalculatorService>(svc => {
        svc.AddEndpoint<ICalculator, NetTcpBinding>( ... );
        svc.AddEndpoint<ICalculator, AnotherBinding>( ... );
    });
});
```

Probably I miss some use case that can't be solved this way?

(for context, a couple of years ago I wrote a .NET Standard wrapper for WCF services hosted in .NET Framework, [here is a short read up](https://github.com/emgdev/dotnet-utils/tree/master/docs/libraries/ServiceModel))

**Comment by mconnew** [comment]
@Kralizek, I just notived your question with regards the api shape. It basically comes down to the fact that I designed the api working on this project on my own with zero external feedback. I created something which works and mirrors the self hosted WCF code reasonably well with the idea to come back and improve it later. Then I got approval to release my prototype and it never got improved as it worked well enough. We just discussed this in our community meeting this morning and we're going to have an open public discussion about changing it to something better for the 2.0 release. Basically the current plan is to release WSDL and WS-Federation support and move to a 1.0 version. Then start working on 2.0 where we drop .NET Framework support and move to .NET 6, and make api changes like improve the configuration api's.
